File: data/snippets/quiz017.cpp
// 为可能的未来需求而过度设计的日志系统
class AbstractLogStrategy {
public:
    virtual void log(const std::string& message, int level) = 0;
    virtual void initialize() {}
    virtual void cleanup() {}
    virtual bool isEnabled() { return true; }
    virtual void setFormat(const std::string& format) {}
    virtual std::string getLastLog() { return ""; }
    virtual ~AbstractLogStrategy() = default;
};

class LogManager {
private:
    std::vector<std::shared_ptr<AbstractLogStrategy>> strategies;
    std::map<std::string, int> logLevels;
    std::map<std::string, std::function<void(const std::string&)>> customHandlers;
    std::vector<std::string> logBuffer;
    bool enableBuffering;
    int maxBufferSize;
    std::string defaultFormat;
    
public:
    LogManager() : enableBuffering(false), maxBufferSize(100) {
        // 初始化默认日志级别
        logLevels["DEBUG"] = 0;
        logLevels["INFO"] = 1;
        logLevels["WARNING"] = 2;
        logLevels["ERROR"] = 3;
        logLevels["CRITICAL"] = 4;
        
        defaultFormat = "[%time%] [%level%] %message%";
    }
    
    void addStrategy(std::shared_ptr<AbstractLogStrategy> strategy) {
        strategies.push_back(strategy);
    }
    
    void setBuffering(bool enable, int size = 100) {
        enableBuffering = enable;
        maxBufferSize = size;
    }
    
    void addCustomHandler(const std::string& name, 
                         std::function<void(const std::string&)> handler) {
        customHandlers[name] = handler;
    }
    
    void log(const std::string& message, const std::string& level = "INFO") {
        int logLevel = logLevels[level];
        
        if (enableBuffering) {
            logBuffer.push_back(message);
            if (logBuffer.size() > maxBufferSize) {
                logBuffer.erase(logBuffer.begin());
            }
        }
        
        for (auto& strategy : strategies) {
            if (strategy->isEnabled()) {
                strategy->log(message, logLevel);
            }
        }
        
        for (auto& [name, handler] : customHandlers) {
            handler(message);
        }
    }
    
    void flushBuffer() {
        logBuffer.clear();
    }
};

// 实际应用中只使用了这个简单的控制台日志实现
class ConsoleLogStrategy : public AbstractLogStrategy {
public:
    void log(const std::string& message, int level) override {
        std::cout << "LOG [Level " << level << "]: " << message << std::endl;
    }
};

// 客户端代码使用日志系统
void clientCode() {
    LogManager logManager;
    auto consoleLog = std::make_shared<ConsoleLogStrategy>();
    logManager.addStrategy(consoleLog);
    
    // 实际上，客户端只使用了基本的日志功能
    logManager.log("Application started", "INFO");
    logManager.log("Operation completed", "INFO");
}

File: data/snippets/quiz017.java
import java.util.*;
import java.util.function.Consumer;

abstract class AbstractLogStrategy {
    public abstract void log(String message, int level);
    public void initialize() {}
    public void cleanup() {}
    public boolean isEnabled() { return true; }
    public void setFormat(String format) {}
    public String getLastLog() { return ""; }
    public void setThreshold(int threshold) {}
    public Map<String, Object> getMetrics() { return new HashMap<>(); }
}

interface LogFilter {
    boolean shouldLog(String message, int level);
}

interface LogFormatter {
    String format(String message, int level, Date timestamp);
}

class LogManager {
    private List<AbstractLogStrategy> strategies;
    private Map<String, Integer> logLevels;
    private Map<String, Consumer<String>> customHandlers;
    private List<String> logBuffer;
    private boolean enableBuffering;
    private int maxBufferSize;
    private String defaultFormat;
    private List<LogFilter> filters;
    private LogFormatter formatter;
    private boolean enableAsync;
    private int asyncQueueSize;
    private Map<String, Object> configuration;
    
    public LogManager() {
        this.strategies = new ArrayList<>();
        this.logLevels = new HashMap<>();
        this.customHandlers = new HashMap<>();
        this.logBuffer = new ArrayList<>();
        this.enableBuffering = false;
        this.maxBufferSize = 100;
        this.filters = new ArrayList<>();
        this.enableAsync = false;
        this.asyncQueueSize = 1000;
        this.configuration = new HashMap<>();
        
        initializeLogLevels();
        this.defaultFormat = "[%time%] [%level%] %message%";
    }
    
    private void initializeLogLevels() {
        logLevels.put("DEBUG", 0);
        logLevels.put("INFO", 1);
        logLevels.put("WARNING", 2);
        logLevels.put("ERROR", 3);
        logLevels.put("CRITICAL", 4);
    }
    
    public void addStrategy(AbstractLogStrategy strategy) {
        strategies.add(strategy);
    }
    
    public void removeStrategy(AbstractLogStrategy strategy) {
        strategies.remove(strategy);
    }
    
    public void setBuffering(boolean enable, int size) {
        this.enableBuffering = enable;
        this.maxBufferSize = size;
    }
    
    public void addCustomHandler(String name, Consumer<String> handler) {
        customHandlers.put(name, handler);
    }
    
    public void removeCustomHandler(String name) {
        customHandlers.remove(name);
    }
    
    public void addFilter(LogFilter filter) {
        filters.add(filter);
    }
    
    public void setFormatter(LogFormatter formatter) {
        this.formatter = formatter;
    }
    
    public void setAsyncMode(boolean enable, int queueSize) {
        this.enableAsync = enable;
        this.asyncQueueSize = queueSize;
    }
    
    public void setConfiguration(String key, Object value) {
        configuration.put(key, value);
    }
    
    public Object getConfiguration(String key) {
        return configuration.get(key);
    }
    
    public void log(String message, String level) {
        Integer logLevel = logLevels.get(level);
        if (logLevel == null) logLevel = 1;
        
        for (LogFilter filter : filters) {
            if (!filter.shouldLog(message, logLevel)) {
                return;
            }
        }
        
        if (enableBuffering) {
            logBuffer.add(message);
            if (logBuffer.size() > maxBufferSize) {
                logBuffer.remove(0);
            }
        }
        
        for (AbstractLogStrategy strategy : strategies) {
            if (strategy.isEnabled()) {
                strategy.log(message, logLevel);
            }
        }
        
        for (Consumer<String> handler : customHandlers.values()) {
            handler.accept(message);
        }
    }
    
    public void log(String message) {
        log(message, "INFO");
    }
    
    public void flushBuffer() {
        logBuffer.clear();
    }
    
    public List<String> getBufferedLogs() {
        return new ArrayList<>(logBuffer);
    }
    
    public void shutdown() {
        for (AbstractLogStrategy strategy : strategies) {
            strategy.cleanup();
        }
    }
    
    public Map<String, Object> getSystemMetrics() {
        Map<String, Object> metrics = new HashMap<>();
        metrics.put("bufferSize", logBuffer.size());
        metrics.put("strategiesCount", strategies.size());
        metrics.put("handlersCount", customHandlers.size());
        return metrics;
    }
}

// In practice, only this simple console log implementation is used
class ConsoleLogStrategy extends AbstractLogStrategy {
    @Override
    public void log(String message, int level) {
        System.out.println("LOG [Level " + level + "]: " + message);
    }
}

class ClientCode {
    public void demonstrateUsage() {
        LogManager logManager = new LogManager();
        ConsoleLogStrategy consoleLog = new ConsoleLogStrategy();
        logManager.addStrategy(consoleLog);
        
        logManager.log("Application started", "INFO");
        logManager.log("Operation completed", "INFO");
    }
}
